### PR TITLE
[MAINTENANCE] Enable `B009` Bugbear `get-attr-with-constant` rule

### PR DESCRIPTION
**Pull request by Kilo59**
Enable the B009 rule and fix existing violations.
https://beta.ruff.rs/docs/rules/get-attr-with-constant/#what-it-does

We've already adopted the `Bugbear` rules, but we initially excluded a couple of rules that had a good number of violations.
  - https://github.com/great-expectations/great_expectations/pull/7776#discussion_r1180870711

> ### Why this is bad?
> `getattr` is used to access attributes dynamically. If the attribute is defined as a constant, it is no safer than a typical property access. When possible, prefer property access over `getattr` calls, as the former is more concise and idiomatic.

```python
getattr(obj, "foo")  # without a default this will still cause an Attribute error
# should be re-written as...
obj.foo
# or
getattr(obj, "foo", SOME_DEFAULT)
```

## Questions for Reviewers
Should any of these changes instead be updated to use a default in the form -> `getattr(obj, "foo", None)`?

These changes won't introduce any new bugs for the same reason this linting rule exists.
But this PR provides an opportunity to potentially address some latent, existing bugs.

## TODO

- [x] Rewrite as `getattr(obj, "foo", None)` where needed

------

- [x] Description of PR changes above includes a link to [an existing GitHub issue](https://github.com/great-expectations/great_expectations/issues)
- [x] PR title is prefixed with one of: [BUGFIX], [FEATURE], [DOCS], [MAINTENANCE], [CONTRIB]
- [x] Code is linted - run `invoke lint` (uses `black` + `ruff`)
- [x] Appropriate tests and docs have been updated

For more details, see our [Contribution Checklist](https://docs.greatexpectations.io/docs/contributing/contributing_checklist), [Coding style guide](https://docs.greatexpectations.io/docs/contributing/style_guides/code_style), and [Documentation style guide](https://docs.greatexpectations.io/docs/contributing/style_guides/docs_style).

After you submit your PR, keep the page open and **monitor the statuses of the various checks made by our continuous integration process at the bottom of the page. Please fix any issues that come up** and [reach out on Slack](https://greatexpectations.io/slack) if you need help. Thanks for contributing!

### CODE REVIEW

File: great_expectations/data_context/data_context/abstract_data_context.py [hunk 1]
```diff
@@ -2662,7 +2662,7 @@ def get_validator(  # noqa: PLR0913
             > 1
         ):
             ge_cloud_mode = getattr(  # attr not on AbstractDataContext
-                self, "ge_cloud_mode"
```
Comment (Kilo59) [inline great_expectations/data_context/data_context/abstract_data_context.py]: More thoroughly addressed in this PR
* https://github.com/great-expectations/great_expectations/pull/8433

File: great_expectations/expectations/metrics/map_metric_provider/map_metric_provider.py [hunk 2]
```diff
@@ -114,7 +114,7 @@ def _register_metric_functions(cls):  # noqa: C901, PLR0912, PLR0915
                 # This is not a metric.
                 continue
 
-            metric_fn_type = getattr(candidate_metric_fn, "metric_fn_type")
+            metric_fn_type = candidate_metric_fn.metric_fn_type
```
Comment (Kilo59) [inline great_expectations/expectations/metrics/map_metric_provider/map_metric_provider.py:117]:
```suggestion
            metric_fn_type = getattr(candidate_metric_fn, "metric_fn_type", None)
```

File: great_expectations/expectations/metrics/meta_metric_provider.py [hunk 3]
```diff
@@ -36,7 +36,7 @@ def __new__(cls, name, bases, classdict, *args, **kwargs):
         if alias is not None:
 
             def new(cls, *args, **kwargs):
-                alias = getattr(cls, "_DeprecatedMetaMetricProvider__alias")
+                alias = cls._DeprecatedMetaMetricProvider__alias
```
Comment (Kilo59) [inline great_expectations/expectations/metrics/meta_metric_provider.py:39]: Per the deprecation warning below it looks like this and the deprecation warning can just be removed.

File: great_expectations/expectations/set_based_column_map_expectation.py
Comment (Kilo59) [inline great_expectations/expectations/set_based_column_map_expectation.py:1]: The logic here seems a bit suspect.

File: great_expectations/expectations/regex_based_column_map_expectation.py [hunk 4]
```diff
@@ -187,7 +187,7 @@ def validate_configuration(
     def _question_renderer(cls, configuration, result=None, runtime_configuration=None):
         column = configuration.kwargs.get("column")
         mostly = configuration.kwargs.get("mostly")
-        regex = getattr(cls, "regex")
+        regex = cls.regex
```
Comment (Kilo59) [inline great_expectations/expectations/regex_based_column_map_expectation.py]:
```suggestion
        regex = getattr(cls, "regex", None)
```
Not sure about this one...

File: great_expectations/expectations/metrics/metric_provider.py [hunk 5]
```diff
@@ -153,8 +153,7 @@ def _register_metric_functions(cls) -> None:
                 # This is not a metric or renderer.
                 continue
 
-            if hasattr(attr_obj, "metric_engine"):
-                engine = getattr(attr_obj, "metric_engine")
+            if engine := getattr(attr_obj, "metric_engine", None):
```
Comment (tyler-hoffman) [inline great_expectations/expectations/metrics/metric_provider.py:156]: ❤️

**Review by cdkini — APPROVED** [review]

**Pull request merged**